src/libstd/sys/unix/fs.rs

```diff
@@ -651,6 +658,13 @@ fn remove_dir_all_recursive(path: &Path) -> io::Result<()> {
     for child in try!(readdir(path)) {
         let child = try!(child);
         if try!(child.file_type()).is_dir() {
+            let mut mode = try!(child.metadata()).perm();
+            if mode.readonly() {
+                // we can only remove the contents of a directory if it is not
+                // readonly
+                mode.set_readonly(false);
+                try!(set_perm(&child.path(), mode));
+            }
```

Comment:
Perhaps this function could take `&Metadata` as an argument and perform this logic at the start? That way it could prevent duplication between here/above.